File: apina-core/src/main/java/fi/evident/apina/java/model/type/TypeSchema.java
package fi.evident.apina.java.model.type;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static java.util.Collections.emptyList;
import static java.util.Collections.unmodifiableList;

public final class TypeSchema {

    private final List<JavaTypeVariable> variables = new ArrayList<>();
    private final Map<JavaTypeVariable, List<JavaType>> boundMap = new HashMap<>();

    public void add(JavaTypeVariable var) {
        List<JavaType> old = boundMap.putIfAbsent(var, new ArrayList<>());
        if (old != null)
            throw new IllegalArgumentException("tried to add duplicate variable: " + var);
        variables.add(var);
    }

    public void addBound(JavaTypeVariable var, JavaType bound) {
        getBoundList(var).add(bound);
    }

    public List<JavaType> getTypeBounds(JavaTypeVariable var) {
        return unmodifiableList(getBoundList(var));
    }

    public List<JavaTypeVariable> getVariables() {
        return unmodifiableList(variables);
    }

    @NotNull
    private List<JavaType> getBoundList(JavaTypeVariable var) {
        List<JavaType> bounds = boundMap.get(var);
        if (bounds != null)
            return bounds;
        else
            return emptyList();
    }

    @Override
    public String toString() {
        return boundMap.toString();
    }

    public boolean isEmpty() {
        return boundMap.isEmpty();
    }

    @NotNull
    public TypeEnvironment apply(List<JavaType> arguments) {
        if (arguments.size() != variables.size())
            throw new IllegalArgumentException("expected " + variables.size() + " arguments, but got " + arguments.size());

        TypeEnvironment env = TypeEnvironment.empty();
        int index = 0;
        for (JavaTypeVariable var : variables)
            env.bind(var, arguments.get(index++));

        return env;
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/model/type/TypeSchema.kt
package fi.evident.apina.java.model.type

import java.util.*

class TypeSchema {

    private val _variables = ArrayList<JavaType.Variable>()
    private val boundMap = HashMap<JavaType.Variable, MutableList<JavaType>>()

    val variables: List<JavaType.Variable>
        get() = _variables

    fun add(v: JavaType.Variable) {
        val old = boundMap.putIfAbsent(v, ArrayList())
        if (old != null)
            throw IllegalArgumentException("tried to add duplicate variable: $v")
        _variables.add(v)
    }

    fun addBound(v: JavaType.Variable, bound: JavaType) {
        val bounds = boundMap[v] ?: throw IllegalArgumentException("unknown variable $v")
        bounds.add(bound)
    }

    fun getTypeBounds(v: JavaType.Variable): List<JavaType> = boundMap[v].orEmpty()

    override fun toString() = boundMap.toString()

    val isEmpty: Boolean
        get() = boundMap.isEmpty()

    fun apply(arguments: List<JavaType>): TypeEnvironment {
        require(arguments.size == variables.size) { "expected ${variables.size} arguments, but got ${arguments.size}" }

        val env = TypeEnvironment.empty()

        variables.forEachIndexed { i, v ->
            env[v] = arguments[i]
        }

        return env
    }
}
